src/Widgets/NotificationsList.vala

```diff
@@ -129,11 +134,23 @@ public class Notifications.NotificationsList : Gtk.ListBox {
 
     private void on_row_activated (Gtk.ListBoxRow row) {
         if (row is NotificationEntry) {
-            unowned NotificationEntry notification_entry = (NotificationEntry) row;
-            notification_entry.notification.run_default_action ();
-            notification_entry.clear ();
+            var notification_entry = (NotificationEntry) row;
+
+            if (notification_entry.notification.default_action != null) {
+                unowned var action_group = get_action_group (ACTION_GROUP_PREFIX);
+                action_group.activate_action (notification_entry.notification.default_action, null);
+            } else if (notification_entry.notification.actions.length > 0) {
+                return;
+            }
+
+            try {
+                notification_entry.notification.app_info.launch (null, null);
 
-            close_popover ();
+                notification_entry.clear ();
+                close_popover ();
+            } catch (Error e) {
+                warning ("Unable to launch app: %s", e.message);
+            }
```

Comment:
this should be in a else block and close_popover() should be called after it.

Reply:
But if it's in a else block it can't do

> It only launches the AppInfo now without an action. I think that's something we want or shouldn't we launch it at all? IMHO it's probably expected to bring the app to foreground when the notification is clicked on regardless of default action or no action.

right? Do we still want it?

Reply:
We want to call `GLib.AppInfo.launch()` only when there's no default action, the current code always call it.